File: src/DesktopGrass.Win2D/MouseHook.cs
// MouseHook.cs - WH_MOUSE_LL global hook + bounded SPSC event channel.
//
// The low-level mouse hook runs synchronously on the hook thread (the
// thread that installed the hook, in our case the main UI thread). The
// callback must:
//   * Always return CallNextHookEx so we never consume input.
//   * Do minimal work — queue an event and return.
//
// We use System.Threading.Channels with a bounded capacity and DropOldest
// full-mode so a runaway producer (rapid cursor motion) can't bloat memory.
// The renderer drains the channel synchronously each frame.

using System;
using System.Threading.Channels;
using DesktopGrass.Win2D.Interop;

namespace DesktopGrass.Win2D;

internal enum HookEventKind { Move, LeftClick }

internal readonly struct HookEvent
{
    public readonly HookEventKind Kind;
    public readonly int ScreenX;
    public readonly int ScreenY;
    public readonly double Time;

    public HookEvent(HookEventKind kind, int x, int y, double time)
    {
        Kind = kind; ScreenX = x; ScreenY = y; Time = time;
    }
}

internal sealed class MouseHook : IDisposable
{
    // The delegate MUST be a static field (or instance field rooted by a
    // long-lived owner) so the GC doesn't collect it while the hook is
    // installed. A captured local would be reclaimed and the hook would
    // crash on the next mouse event.
    private readonly Win32.HookProc _proc;
    private IntPtr _hook;
    private readonly Channel<HookEvent> _channel;
    private readonly long _qpcFreq;

    public ChannelReader<HookEvent> Reader => _channel.Reader;

    public MouseHook()
    {
        _channel = Channel.CreateBounded<HookEvent>(new BoundedChannelOptions(capacity: 1024)
        {
            SingleReader = true,
            SingleWriter = true,
            FullMode = BoundedChannelFullMode.DropOldest,
        });
        _proc = HookCallback;
        Win32.QueryPerformanceFrequency(out _qpcFreq);
    }

    public void Install()
    {
        // hMod=0 with a managed callback is documented as supported when
        // the hook is installed for the calling thread's process only via
        // LowLevelMouseProc; in practice WH_MOUSE_LL requires the module
        // handle of the calling process so its code is callable by the
        // hook thread. Passing GetModuleHandle(null) is the standard idiom.
        var hMod = Win32.GetModuleHandleW(null);
        _hook = Win32.SetWindowsHookExW(Win32.WH_MOUSE_LL, _proc, hMod, 0);
        if (_hook == IntPtr.Zero)
        {
            int err = System.Runtime.InteropServices.Marshal.GetLastWin32Error();
            throw new InvalidOperationException($"SetWindowsHookEx WH_MOUSE_LL failed: error {err}");
        }
    }

    private IntPtr HookCallback(int nCode, IntPtr wParam, IntPtr lParam)
    {
        if (nCode >= 0)
        {
            uint msg = (uint)wParam.ToInt32();
            var data = System.Runtime.InteropServices.Marshal.PtrToStructure<Win32.MSLLHOOKSTRUCT>(lParam);
            double t = NowSeconds();
            switch (msg)
            {
                case Win32.WM_MOUSEMOVE:
                    _channel.Writer.TryWrite(new HookEvent(HookEventKind.Move, data.Pt.X, data.Pt.Y, t));
                    break;
                case Win32.WM_LBUTTONDOWN:
                    _channel.Writer.TryWrite(new HookEvent(HookEventKind.LeftClick, data.Pt.X, data.Pt.Y, t));
                    break;
            }
        }
        return Win32.CallNextHookEx(_hook, nCode, wParam, lParam);
    }

    private double NowSeconds()
    {
        Win32.QueryPerformanceCounter(out long now);
        return (double)now / _qpcFreq;
    }

    public void Dispose()
    {
        if (_hook != IntPtr.Zero)
        {
            Win32.UnhookWindowsHookEx(_hook);
            _hook = IntPtr.Zero;
        }
        _channel.Writer.TryComplete();
    }
}

File: src/DesktopGrass.Native/src/MouseHook.cpp
// MouseHook.cpp

#include "MouseHook.h"

#include <atomic>
#include <chrono>

namespace desktopgrass {

namespace {

std::atomic<MouseEventQueue*> g_queue{nullptr};
std::atomic<bool>             g_registrationActive{false};
LARGE_INTEGER                 g_qpcFreq{};
LARGE_INTEGER                 g_qpcStart{};

HHOOK install_platform_hook(HOOKPROC callback) noexcept {
    return SetWindowsHookExW(
        WH_MOUSE_LL, callback, GetModuleHandleW(nullptr), 0);
}

BOOL uninstall_platform_hook(HHOOK hook) noexcept {
    return UnhookWindowsHookEx(hook);
}

constexpr MouseHookPlatform kDefaultPlatform{
    install_platform_hook,
    uninstall_platform_hook,
};

double now_seconds() noexcept {
    LARGE_INTEGER c;
    QueryPerformanceCounter(&c);
    return static_cast<double>(c.QuadPart - g_qpcStart.QuadPart) /
           static_cast<double>(g_qpcFreq.QuadPart);
}

LRESULT CALLBACK LowLevelMouseProc(int nCode, WPARAM wParam, LPARAM lParam) {
    // Per spec: always pass the event through. Never consume.
    if (nCode == HC_ACTION) {
        MouseEventQueue* q = g_queue.load(std::memory_order_acquire);
        if (q) {
            const MSLLHOOKSTRUCT* m = reinterpret_cast<const MSLLHOOKSTRUCT*>(lParam);
            RawMouseEvent ev{};
            ev.timeSeconds = now_seconds();
            ev.screenX     = m->pt.x;
            ev.screenY     = m->pt.y;

            switch (wParam) {
                case WM_MOUSEMOVE:
                    ev.type = EventType::Move;
                    q->push(ev);
                    break;
                case WM_LBUTTONDOWN:
                    ev.type = EventType::Click;
                    q->push(ev);
                    break;
                default:
                    break;
            }
        }
    }
    return CallNextHookEx(nullptr, nCode, wParam, lParam);
}

} // anonymous

MouseHookRegistration::MouseHookRegistration() noexcept
    : platform_(kDefaultPlatform) {}

MouseHookRegistration::MouseHookRegistration(
    MouseHookPlatform platform) noexcept
    : platform_(platform) {}

MouseHookRegistration::~MouseHookRegistration() {
    Reset();
}

bool MouseHookRegistration::Install(MouseEventQueue* queue) noexcept {
    if (!queue || hook_ || !platform_.install || !platform_.uninstall) {
        return false;
    }

    bool expected = false;
    if (!g_registrationActive.compare_exchange_strong(
            expected, true, std::memory_order_acq_rel)) {
        return false;
    }

    if (g_qpcFreq.QuadPart == 0) {
        QueryPerformanceFrequency(&g_qpcFreq);
        QueryPerformanceCounter(&g_qpcStart);
    }

    g_queue.store(queue, std::memory_order_release);
    hook_ = platform_.install(LowLevelMouseProc);
    if (!hook_) {
        g_queue.store(nullptr, std::memory_order_release);
        g_registrationActive.store(false, std::memory_order_release);
        return false;
    }
    return true;
}

void MouseHookRegistration::Reset() noexcept {
    if (!hook_) return;

    g_queue.store(nullptr, std::memory_order_release);
    platform_.uninstall(hook_);
    hook_ = nullptr;
    g_registrationActive.store(false, std::memory_order_release);
}

} // namespace desktopgrass
